packages/shared_preferences/shared_preferences_android/android/src/main/java/io/flutter/plugins/sharedpreferences/SharedPreferencesPlugin.java

```diff
@@ -5,42 +5,207 @@
+      Log.e(TAG, "Received exception while setting up SharedPreferencesPlugin", ex);
+    }
   }
 
   @Override
-  public void onAttachedToEngine(FlutterPlugin.FlutterPluginBinding binding) {
-    setupChannel(binding.getBinaryMessenger(), binding.getApplicationContext());
+  public void onAttachedToEngine(@NonNull FlutterPlugin.FlutterPluginBinding binding) {
+    setUp(binding.getBinaryMessenger(), binding.getApplicationContext());
   }
 
   @Override
-  public void onDetachedFromEngine(FlutterPlugin.FlutterPluginBinding binding) {
-    teardownChannel();
+  public void onDetachedFromEngine(@NonNull FlutterPlugin.FlutterPluginBinding binding) {
+    SharedPreferencesApi.setup(binding.getBinaryMessenger(), null);
   }
 
-  private void setupChannel(BinaryMessenger messenger, Context context) {
-    channel = new MethodChannel(messenger, CHANNEL_NAME);
-    handler = new MethodCallHandlerImpl(context);
-    channel.setMethodCallHandler(handler);
+  @Override
+  public @NonNull Boolean setBool(@NonNull String key, @NonNull Boolean value) {
+    return preferences.edit().putBoolean(key, value).commit();
+  }
+
+  @Override
+  public @NonNull Boolean setString(@NonNull String key, @NonNull String value) {
+    // TODO (tarrinneal): Move this string prefix checking logic to dart code and make it an Argument Error.
+    if (value.startsWith(LIST_IDENTIFIER)
+        || value.startsWith(BIG_INTEGER_PREFIX)
+        || value.startsWith(DOUBLE_PREFIX)) {
+      throw new RuntimeException(
+          "StorageError: This string cannot be stored as it clashes with special identifier prefixes");
+    }
+    return preferences.edit().putString(key, value).commit();
+  }
+
+  @Override
+  public @NonNull Boolean setInt(@NonNull String key, @NonNull Long value) {
+    return preferences.edit().putLong(key, value).commit();
+  }
+
+  @Override
+  public @NonNull Boolean setDouble(@NonNull String key, @NonNull Double value) {
+    String doubleValueStr = Double.toString(value);
+    return preferences.edit().putString(key, DOUBLE_PREFIX + doubleValueStr).commit();
   }
 
-  private void teardownChannel() {
-    handler.teardown();
-    handler = null;
-    channel.setMethodCallHandler(null);
-    channel = null;
+  @Override
+  public @NonNull Boolean remove(@NonNull String key) {
+    return preferences.edit().remove(key).commit();
+  }
+
+  @Override
+  public @NonNull Boolean setStringList(@NonNull String key, @NonNull List<String> value)
+      throws RuntimeException {
+    return preferences.edit().putString(key, LIST_IDENTIFIER + listEncoder.encode(value)).commit();
+  }
+
+  @Override
+  public @NonNull Map<String, Object> getAllWithPrefix(@NonNull String prefix)
+      throws RuntimeException {
+    return getAllPrefs(prefix);
+  }
+
+  @Override
+  public @NonNull Boolean clearWithPrefix(@NonNull String prefix) throws RuntimeException {
+    Map<String, ?> allPrefs = preferences.getAll();
+    Map<String, Object> filteredPrefs = new HashMap<>();
```

Comment:
This doesn't need to be a `Map`, just a `List<String>` of keys. You're never using the value when clearing.